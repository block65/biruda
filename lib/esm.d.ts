interface ImportMeta {
  resolve: (specifier: string, parent?: string) => Promise<string>;
}
