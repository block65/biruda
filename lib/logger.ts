import { createCliLogger } from '@block65/logger';

export const logger = createCliLogger({
  level: 'info',
  // pretty: true,
});

logger.on('level-change', (lvl, val, prevLvl, prevVal) => {
  if (lvl !== prevLvl) {
    logger.info(
      'Logger level %s (%d) was changed to %s (%d)',
      prevLvl,
      prevVal,
      lvl,
      val,
    );
  }
});
