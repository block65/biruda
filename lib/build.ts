import type { PackageJson, TsConfigJson } from 'type-fest';
import * as esbuild from 'esbuild';
import { lstatSync, statSync, writeFileSync } from 'fs';
import { dir } from 'tmp-promise';
import pkgUp from 'pkg-up';
import { dirname, resolve } from 'path';
import { logger } from './logger';

// all optional because they might be on command like
export interface BirudaConfigFileProperties {
  entryPoints?: string[];
  verbose?: boolean;
  outDir?: string;
  platform?: string;
  externals?: string[];
  forceInclude?: string[];
  ignorePackages?: string[];
  archiveFormat?: 'zip' | 'tar';
  sourceMapSupport?: boolean;
}

export interface BirudaCliArguments {
  config?: string;
  verbose?: boolean;
  output?: string;
  // baseDir?: string;
  entrypoint?: string[];
  archiveFormat?: string;
  forceInclude?: string[];
  sourceMapSupport?: boolean;
}

export interface BirudaBuildOptions {
  verbose?: boolean;
  outDir: string;
  entryPoint: string;
  // baseDir: string;
  platform: string;
  externals?: string[];
  ignorePackages?: string[];
  forceInclude?: string[];
  sourceMapSupport?: boolean;
}

export async function build(
  options: BirudaBuildOptions,
): Promise<{ tmpFile: string; tmpDir: string; cleanup: () => Promise<void> }> {
  const { entryPoint } = options;

  const entryPointStat = statSync(entryPoint);
  if (!entryPointStat.isFile()) {
    throw new Error(`Invalid entrypoint ${entryPoint}`);
  }

  // eslint-disable-next-line global-require,import/no-dynamic-require
  // const packageJson: PackageJson = require(resolve(baseDir, 'package.json'));
  const packageJsonPath = await pkgUp({
    cwd: dirname(entryPoint),
  });

  if (!packageJsonPath) {
    throw new Error('Cant find a package.json');
  }

  // eslint-disable-next-line global-require,import/no-dynamic-require
  const packageJson: PackageJson = require(packageJsonPath);

  // eslint-disable-next-line global-require,import/no-dynamic-require
  const tsConfigJson: TsConfigJson = require(resolve(
    dirname(packageJsonPath),
    'tsconfig.json',
  ));

  // we need to perform the process inside this directory
  // so we create a tmp file before we move the resulting bundle
  // back out to where the user wanted
  const { path: tmpDir, cleanup } = await dir({
    tmpdir: dirname(entryPoint),
    template: '.tmp-biruda-XXXXXX',
    unsafeCleanup: true,
  });

  // process.on('beforeExit', cleanup);

  const esBuildOutputFilePath = resolve(tmpDir, 'index.js');

  const finalEsBuildOptions: esbuild.BuildOptions = {
    platform: options.platform === 'browser' ? 'browser' : 'node',
    logLevel: options.verbose ? 'info' : 'error',
    external: [...(options.ignorePackages || []), ...(options.externals || [])],
    entryPoints: [entryPoint], // [maybeMakeAbsolute(entryPoint, baseDir)],
    outfile: esBuildOutputFilePath,
    // metafile: '/tmp/meta.json',
    bundle: true,
    minify: false,
    color: true,
    target: tsConfigJson.compilerOptions?.target,
    sourcemap: true, // 'external',
    errorLimit: 1,
    define: {
      NODE_ENV: 'production',
    },
  };

  logger.info('Building entryPoints %s ...', entryPoint);

  return esbuild.build(finalEsBuildOptions).then(async (buildResult) => {
    if (buildResult.warnings.length > 0) {
      logger.warn('Build warnings: %d', buildResult.warnings.length);

      buildResult.warnings.forEach((warn) => {
        logger.trace(warn, `Warning: ${warn.text}`);
      });
    }

    const newPackageJson: PackageJson.PackageJsonStandard = {
      name: packageJson.name,
      version: packageJson.version,
      license: packageJson.license,
      private: packageJson.private,
      // main: path.basename(entryPoint),
      // scripts: Object.fromEntries(
      //   Object.entries(packageJson.scripts || {}).filter(([scriptName]) => {
      //     return !scriptName.match(/^(dev|build|test)(\W|$)/);
      //   }),
      // ),
      // scripts: {
      //   start: 'node index.js',
      // },
      // dependencies: Object.fromEntries(
      //   Array.from(dependencies.entries()).map(([id, { version }]) => [
      //     id,
      //     version,
      //   ]),
      // ),
    };

    const stat = lstatSync(esBuildOutputFilePath);

    if (stat.size === 0) {
      logger.warn(`Empty output file`);
      logger.trace({ stat }, `Stat file %s`, esBuildOutputFilePath);
      throw new Error('Empty output file');
    }

    // const absoluteOutputFile = maybeMakeAbsolute('index.js', tmpDir);
    // copyFileSync(esBuildOutputFilePath, absoluteOutputFile);
    // logger.warn(
    //   {
    //     stat: lstatSync(esBuildOutputFilePath),
    //     esBuildOutputFilePath,
    //     absoluteOutputFile,
    //   },
    //   `Stat file %s after:copyFileSync0`,
    // );
    // copyFileSync(`${esBuildOutputFilePath}.map`, `${absoluteOutputFile}.map`);

    writeFileSync(
      resolve(tmpDir, 'package.json'),
      JSON.stringify(newPackageJson, null, 2),
    );
    logger.info(`Build completed. Output is in ${tmpDir}`);

    return { tmpFile: esBuildOutputFilePath, tmpDir, cleanup };
  });
  // .catch((err) => logger.error(err));
}
