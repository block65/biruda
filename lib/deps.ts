import { nodeFileTrace, NodeFileTraceReasons } from '@vercel/nft';
import * as fs from 'fs';
import { access, readFile } from 'fs/promises';
import micromatch from 'micromatch';
import pMemoize from 'p-memoize';
import { dirname, normalize, relative } from 'path';
import type { PackageJson } from 'type-fest';
import { fileURLToPath, pathToFileURL, URL } from 'url';
import * as util from 'util';
import { logger as parentLogger } from './logger.js';
import { loadJson, relativeUrl, resolvePackageJson } from './utils.js';

const logger = parentLogger.child({ name: 'deps' });

export interface RecursiveDependency extends Dependency {
  deps: RecursiveDependency[];
}

export interface Dependency {
  name: string;
  version: string;
  files?: string[];
  path: URL;
}

interface Warning extends Error {
  lineText?: string;
  file?: string;
  line?: number;
  column?: number;
}

async function readManifestInner(
  dirOrFile: string | URL,
  throwOnMissing?: false,
): Promise<PackageJson | null>;
async function readManifestInner(
  dirOrFile: string | URL,
  throwOnMissing?: true,
): Promise<PackageJson>;
async function readManifestInner(
  dirOrFile: string | URL,
  throwOnMissing?: boolean,
): Promise<PackageJson | null> {
  const file = dirOrFile.toString().endsWith('package.json')
    ? dirOrFile
    : new URL('package.json', pathToFileURL(`${dirOrFile.toString()}/`));

  const exists = await access(file, fs.constants.F_OK)
    .then(() => true)
    .catch(() => false);

  if (throwOnMissing || exists) {
    return JSON.parse(await readFile(file, 'utf-8'));
  }
  return null;
}

export const readManifest = pMemoize(readManifestInner, {
  cacheKey: ([fdirOrFile]) => fdirOrFile,
});

export async function findWorkspaceRoot(initial = process.cwd()) {
  logger.trace('Finding workspace root from %s', initial);

  let previousDirectory = null;
  let currentDirectory = normalize(initial);

  do {
    // suppress eslint here because this needs to be sequential/ serial
    // and cannot be parallelised
    // eslint-disable-next-line no-await-in-loop
    const manifest = await readManifest(currentDirectory);

    if (manifest) {
      const workspaces = Array.isArray(manifest.workspaces)
        ? manifest.workspaces
        : manifest.workspaces?.packages;

      if (workspaces) {
        logger.trace(
          { workspaces },
          'Found workspaces in %s',
          currentDirectory,
        );

        const relativePath = relative(currentDirectory, initial);
        if (
          relativePath === '' ||
          micromatch([relativePath], workspaces, { bash: true }).length > 0
        ) {
          logger.trace(
            { patterns: workspaces },
            'Using workspace root at %s',
            currentDirectory,
          );

          return currentDirectory;
        }

        logger.trace(
          { relativePath, patterns: workspaces },
          'Workspace doesnt include me %s',
          currentDirectory,
        );

        return null;
      }
      logger.trace('No workspaces in %s', currentDirectory);
    } else {
      logger.trace('No manifest in %s', currentDirectory);
    }

    previousDirectory = currentDirectory;
    currentDirectory = dirname(currentDirectory);
  } while (currentDirectory !== previousDirectory);

  return null;
}

function flattenDeps(initialChildren: RecursiveDependency[]): Dependency[] {
  // console.log({ children });
  return initialChildren.flatMap((child): Dependency[] => {
    const { deps, ...childWithoutChildren } = child;
    // console.log({ child, flat: flattenDeps(child.children) });
    return deps.length > 0
      ? [childWithoutChildren, ...flattenDeps(deps)]
      : [childWithoutChildren];
  });
}

export async function traceDependencies(
  initialDeps: Record<string, string>,
  startPath: URL,
  // mode = 'production',
): Promise<Dependency[]> {
  const circuitBreaker = new Set<string>();

  function recursiveResolveDependencies(
    deps: Record<string, string>,
    base: URL,
  ): Promise<RecursiveDependency[]> {
    return Promise.all(
      Object.entries(deps).map(
        async ([name, version]): Promise<RecursiveDependency> => {
          const modulePackageJsonUrl = await resolvePackageJson(name, base);
          const packageJson = await loadJson<PackageJson>(modulePackageJsonUrl);

          const nextPath = new URL('.', modulePackageJsonUrl);
          const beenHere = circuitBreaker.has(nextPath.toString());

          // if we've already been here, we just return empty children as
          // a previous iteration has already returned the children
          if (beenHere) {
            return {
              name,
              version,
              path: nextPath,
              files: packageJson.files,
              deps: [],
            };
          }

          circuitBreaker.add(nextPath.toString());

          const children = await recursiveResolveDependencies(
            packageJson.dependencies || {},
            nextPath,
          );

          return {
            name,
            version,
            path: nextPath,
            files: packageJson.files,
            deps: children,
          };
        },
      ),
    );
  }

  return flattenDeps(
    await recursiveResolveDependencies(initialDeps, startPath),
  );
}

export async function traceFiles(
  entryPoints: string[],
  options: {
    baseDir: URL;
    verbose?: boolean;
    ignorePackages?: string[];
  },
): Promise<{
  files: Set<string>;
  base: URL;
  reasons: NodeFileTraceReasons;
}> {
  const { baseDir } = options;

  const maybeWorkspaceRoot = await findWorkspaceRoot(fileURLToPath(baseDir));
  const workspaceRoot = maybeWorkspaceRoot
    ? pathToFileURL(maybeWorkspaceRoot)
    : baseDir;
  // const processCwd = dirname(manifestFilename);

  logger.info(
    entryPoints,
    'Tracing dependencies using base: %s, workspace: %s',
    relativeUrl(workspaceRoot, baseDir),
    workspaceRoot,
  );

  const traceResult = await nodeFileTrace(
    entryPoints.map((entry) => fileURLToPath(new URL(entry, baseDir))),
    {
      base: fileURLToPath(workspaceRoot),
      processCwd: fileURLToPath(baseDir),
      log: options.verbose,
      ignore: options.ignorePackages?.map((pkg) => `node_modules/${pkg}/**`),
      // paths: [base],
      exportsOnly: true,
    },
  );

  const { fileList, esmFileList, reasons } = traceResult;

  if (traceResult.warnings.length > 0) {
    logger.warn('Trace warnings: %d', traceResult.warnings.length);
    traceResult.warnings.forEach((value: Warning) => {
      if (value.lineText) {
        logger.warn(
          { value },
          `${value.message.trim()} caused by ${value.lineText} in ${
            value.file
          }:${value.line}:${value.column}`,
        );
      } else {
        logger.warn(value.message.trim());
      }
    });
  }

  // find and exclude the initial entry points
  const resolvedEntryPoints = Object.entries(reasons)
    .filter(([, reason]) => reason.type === 'initial')
    .map(([file]) => file);

  return {
    base: workspaceRoot,
    reasons,
    files: new Set(
      [...fileList, ...esmFileList].filter(
        (file) => !resolvedEntryPoints.includes(file),
      ),
    ),
  };
}
