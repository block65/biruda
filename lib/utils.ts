import fs from 'fs/promises';
import { createRequire } from 'module';
import { dirname, isAbsolute, relative, resolve } from 'path';
import pkgUp from 'pkg-up';
import type { AsyncReturnType, JsonValue } from 'type-fest';
import { fileURLToPath, pathToFileURL, URL } from 'url';

import { readManifest } from './deps.js';
import { logger as parentLogger } from './logger.js';

const logger = parentLogger.child({ name: 'utils' });

export function maybeMakeAbsolute(entry: string, baseDir: string): string {
  if (isAbsolute(entry)) {
    return entry;
  }
  return resolve(baseDir, entry);
}

export function dedupeArray<T extends any>(arr: T[]): T[] {
  return [...new Set(arr)];
}

// async for future esm compat
export async function resolvePackageJson(
  name: string,
  base: URL,
): Promise<URL> {
  const require = createRequire(base);

  try {
    return pathToFileURL(
      require.resolve(`${name}/package.json`, {
        paths: [base.pathname],
      }),
    );
  } catch (err) {
    if (err.code === 'ERR_PACKAGE_PATH_NOT_EXPORTED') {
      if (logger.isLevelEnabled('debug')) {
        logger.warn({ base, name }, 'WARN: %s', err.message);
      }

      const path = await pkgUp({
        cwd: dirname(
          require.resolve(name, {
            paths: [base.pathname],
          }),
        ),
      });

      if (path === null) {
        throw new Error(`Unable to resolve path $path`);
      }

      return pathToFileURL(path);
    }
    throw err;
  }
}

export async function getDependencyPathsFromModule(
  name: string,
  base: URL,
  descendCallback: (path: URL, name: string) => boolean,
  includeCallback: (path: URL, name: string) => void,
  parents: string[] = [],
): Promise<void> {
  const logPrefixString = [name, parents].join('->');

  logger.debug('[%s] Resolving deps', logPrefixString, name);
  const modulePath = await resolvePackageJson(name, base);

  if (!descendCallback(modulePath, name)) {
    return;
  }

  const pkgJson = await readManifest(modulePath);

  if (!pkgJson) {
    throw new Error(`Unable to locate manifest for ${name}`);
  }

  includeCallback(new URL('package.json', modulePath), name);

  dedupeArray([
    // 'package.json',
    // 'LICENSE',
    // 'LICENSE.md',
    ...(pkgJson.files || []),
  ]).forEach((glob) => {
    includeCallback(new URL(glob, modulePath), name);
  });

  if (pkgJson.main) {
    includeCallback(new URL(pkgJson.main, modulePath), name);
  }

  if (!pkgJson.files) {
    logger.trace('[%s] No files[] in  manifest, adding all', logPrefixString);
    includeCallback(modulePath, name);
  }

  const dependencies = Object.keys(pkgJson.dependencies || {});

  logger.trace(
    '[%s] Found %d deps',
    logPrefixString,
    Object.keys(dependencies).length,
  );

  // at leaf node
  if (dependencies.length === 0) {
    // callback(modulePath);
    return; // [modulePath];
  }

  dependencies.forEach((pkgDep) => {
    // logger.trace('[%s] Found dep %s', pkgDep, name);
    getDependencyPathsFromModule(
      pkgDep,
      modulePath,
      descendCallback,
      includeCallback,
      [...parents, name],
    );
  });
}

export function serialPromiseMapAccum<
  T,
  F extends (arg: T, idx: number, arr: T[]) => Promise<any> = (
    arg: T,
    idx: number,
    arr: T[],
  ) => Promise<any>,
  R = AsyncReturnType<F>
>(arr: T[], fn: F): Promise<R[]> {
  return arr.reduce(async (promise, ...args): Promise<R[]> => {
    const accum = await promise;
    const result = await fn(...args);
    return accum.concat(result);
  }, Promise.resolve([] as R[]));
}

// create a function that will definitely run at least once, every `delay` seconds
export function basicThrottle<T extends (...args: any[]) => any>(
  callback: T,
  delay: number,
) {
  let callable = true;
  return (...args: Parameters<T>): void => {
    if (callable) {
      callable = false;
      setTimeout(() => {
        callable = true;
      }, delay);
      callback(...args);
    }
  };
}

// create a function that will run at least once `delay` seconds after the last call
export function basicDebounce<T extends (...args: any[]) => any>(
  callback: T,
  delay: number,
) {
  let timeout: NodeJS.Timeout;
  return (...args: Parameters<T>) => {
    clearTimeout(timeout);
    timeout = setTimeout(() => callback(...args), delay);
  };
}

export async function loadJson<T = JsonValue>(file: URL): Promise<T> {
  return JSON.parse(await fs.readFile(file, 'utf-8'));
}

export function relativeUrl(from: URL, to: URL): string {
  return relative(fileURLToPath(from), fileURLToPath(to));
}
