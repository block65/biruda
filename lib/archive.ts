import archiver from 'archiver';
import { createWriteStream, existsSync, lstatSync } from 'fs';
import { basename, join, relative, resolve } from 'path';
import { constants } from 'zlib';
import { logger } from './logger';
import { Dependency } from './deps';
import { basicThrottle } from './utils';

function maybeReducePathToNodeModules(path: string, fallback: string): string {
  const regex = /.*?(node_modules.*)/;
  if (path.match(regex)) {
    return path.replace(regex, '$1');
  }
  return `node_modules/${fallback}`;
}

export async function archiveFiles({
  base,
  pkgDir,
  outDir,
  pkgFiles = [],
  files = [],
  extraGlobDirs = [],
  format = 'tar',
}: {
  base: string;
  pkgDir: string;
  outDir: string;
  pkgFiles?: string[];
  files?: string[];
  extraGlobDirs?: string[];
  format?: 'zip' | 'tar';
}) {
  const isTar = format === 'tar';
  const archive = archiver(
    format,
    isTar
      ? {
          gzip: true,
          gzipOptions: {
            level: -1,
          },
        }
      : { zlib: { level: constants.Z_BEST_SPEED } },
  );

  // good practice to catch warnings (ie stat failures and other non-blocking errors)
  archive.on('warning', (err) => {
    if (err.code === 'ENOENT') {
      logger.warn(/* { err }, */ `Archiver warning: ${err.message}`);
    } else {
      // throw error
      throw err;
    }
  });

  // good practice to catch this error explicitly
  archive.on('error', (err) => {
    logger.error(err, `Archiver error: ${err.message}`);
    archive.abort();
    process.exitCode = 1;
    // throw err;
  });

  archive.on(
    'progress',
    basicThrottle((progress) => {
      logger.trace(
        'Progress: %s of %s files archived',
        progress.entries.processed,
        progress.entries.total,
      );
    }, 500),
  );

  const archiveFileName = `pkg.${format}${isTar ? '.gz' : ''}`;
  const output = createWriteStream(`${outDir}/${archiveFileName}`);

  // listen for all archive data to be written
  // 'close' event is fired only when a file descriptor is involved
  output.on('close', () => {
    logger.info(
      `Archive is ~${new Intl.NumberFormat().format(
        Math.round(archive.pointer() / 1024),
      )}kb`,
    );
    logger.trace(
      'Archiver has been finalized and the output file descriptor has closed.',
    );
  });

  // This event is fired when the data source is drained no matter what was the data source.
  // It is not part of this library but rather from the NodeJS Stream API.
  // @see: https://nodejs.org/api/stream.html#stream_event_end
  output.on('end', () => {
    logger.trace('Data has been drained');
  });

  // pipe archive data to the output
  archive.pipe(output);

  logger.trace('Adding pkgDir %s', pkgDir);
  archive.directory(pkgDir, false);

  logger.trace({ pkgFiles }, 'Adding %d extra pkgFiles', pkgFiles.length);

  pkgFiles.forEach((file) => {
    archive.file(join(base, file), {
      name: basename(file),
      // prefix: base,
    });
  });

  logger.trace(
    // { extraGlobDirs },
    'Adding %d extra dirs via glob',
    extraGlobDirs.length,
  );

  extraGlobDirs.forEach((dir) => {
    archive.glob(
      '**/*',
      {
        cwd: dir,
      },
      {
        prefix: relative(base, dir),
      },
    );
  });

  logger.info(`Adding %d files...`, files.length);
  files.forEach((file) => {
    archive.file(join(base, file), {
      name: file,
      // prefix: base,
    });
  });

  logger.trace('Finalizing archive');
  return archive.finalize();
}

export async function archiveDependencies({
  dependencies,
  pkgDir,
  outDir,
  format = 'tar',
}: {
  dependencies: Dependency[];
  pkgDir: string;
  outDir: string;
  format?: 'zip' | 'tar';
}) {
  const isTar = format === 'tar';
  const archive = archiver(
    format,
    isTar
      ? {
          gzip: true,
          gzipOptions: {
            level: -1,
          },
        }
      : { zlib: { level: constants.Z_BEST_SPEED } },
  );

  const archiveFileName = `pkg.${format}${isTar ? '.gz' : ''}`;

  archive.glob('**/*', {
    cwd: pkgDir,
  });

  // good practice to catch warnings (ie stat failures and other non-blocking errors)
  archive.on('warning', (err) => {
    if (err.code === 'ENOENT') {
      logger.warn({ err }, err.message);
    } else {
      // throw error
      throw err;
    }
  });

  // good practice to catch this error explicitly
  archive.on('error', (err) => {
    logger.error(err, 'Archiver error');
    archive.abort();
    process.exitCode = 1;
    // throw err;
  });

  archive.on(
    'progress',
    basicThrottle((progress) => {
      logger.trace(
        'Progress: %s of %s entries archived',
        progress.entries.processed,
        progress.entries.total,
      );
    }, 300),
  );

  const output = createWriteStream(`${outDir}/${archiveFileName}`);

  // listen for all archive data to be written
  // 'close' event is fired only when a file descriptor is involved
  output.on('close', () => {
    logger.info(
      `Archive is ~${new Intl.NumberFormat().format(
        Math.round(archive.pointer() / 1024),
      )}kb`,
    );
    logger.trace(
      'Archiver has been finalized and the output file descriptor has closed.',
    );
  });

  // This event is fired when the data source is drained no matter what was the data source.
  // It is not part of this library but rather from the NodeJS Stream API.
  // @see: https://nodejs.org/api/stream.html#stream_event_end
  output.on('end', () => {
    logger.trace('Data has been drained');
  });

  logger.trace('Piping archiver into output');
  // pipe archive data to the output
  archive.pipe(output);

  logger.trace('Archiving dependencies');

  dependencies.forEach(({ path, name, files, version }) => {
    // const moduleBasePathForArchive = join('node_modules', name);

    const id = `${name}@${version}`;

    if (!files) {
      const destPath = maybeReducePathToNodeModules(path, name);
      logger.trace('[%s] No files[]. Archiving %s into %s', id, path, destPath);
      // archive.directory(path, destPath);
      archive.glob(
        `**/*`,
        {
          cwd: path,
          debug: true,
          ignore: ['node_modules', '*.tsbuildinfo'],
        },
        {
          prefix: destPath,
        },
      );
      return;
    }

    const pkgJsonPrefix = maybeReducePathToNodeModules(path, name);

    // add the package.json
    logger.debug(
      { path },
      '[%s] Archiving package.json into %s',
      id,
      pkgJsonPrefix,
    );

    archive.file(resolve(path, 'package.json'), {
      name: 'package.json',
      prefix: pkgJsonPrefix,
    });

    files
      ?.filter((file) => file !== 'package.json') // we already added it
      .forEach((entry) => {
        // archive.glob(resolve(basePath, file));

        const absoluteFilePath = join(path, entry);
        const looksGlobbish = absoluteFilePath.includes('*');
        const exists = !looksGlobbish && existsSync(absoluteFilePath);

        if (exists) {
          const stats = lstatSync(absoluteFilePath);
          if (stats.isDirectory()) {
            const prefix = maybeReducePathToNodeModules(absoluteFilePath, name);

            logger.debug(
              '[%s] Archiving %s into %s (using dir)',
              id,
              entry,
              prefix,
            );
            // logger.trace({ absoluteFilePath }, 'directory');
            archive.directory(absoluteFilePath, prefix);
          } else if (stats.isFile()) {
            const prefix = maybeReducePathToNodeModules(absoluteFilePath, name);
            // const prefix = `node_modules/${name}`;

            logger.debug(
              '[%s] Archiving %s as %s (using file)',
              id,
              entry,
              prefix,
            );
            // logger.trace({ absoluteFilePath }, 'file');
            archive.file(absoluteFilePath, {
              name: prefix,
            });
          }
        } else {
          if (!looksGlobbish) {
            logger.warn('File %s doesnt exist, trying glob instead', entry);
          }

          const prefix = maybeReducePathToNodeModules(path, name);
          // const prefix = `node_modules/${name}`;

          logger.debug(
            '[%s] Archiving %s into %s (using glob)',
            id,
            entry,
            prefix,
          );
          // logger.trace({ absoluteFilePath, exists, path, entry }, 'glob');

          // if "file" doesnt exist, it might be a glob. Try it.
          archive.glob(
            entry,
            {
              cwd: path,
            },
            {
              prefix,
            },
          );
        }

        // logger.trace({ basePath, file, resolved: resolve(basePath, file) });
      });
  });

  logger.trace('Finalizing archive');
  return archive.finalize();
}
