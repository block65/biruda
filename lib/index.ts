import { loadPackageJson } from './utils.js';

export { loadPackageJson };
