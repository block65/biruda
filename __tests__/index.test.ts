import { PackageJson } from 'type-fest';
import { URL } from 'url';
import { traceDependencies } from '../lib/deps';
import { loadJson } from '../lib/utils.js';

function logNThrow(err: any) {
  console.error(err);
  throw err;
}

describe('Deps Fixtures', () => {
  // test('collect', async () => {});
  // test('flatten', async () => {});

  // test('traceFileDependencies', async () => {
  //   const entryPoint = require.resolve('./fixtures');
  //   // const packageJsonPath = require.resolve('./fixtures/package.json');
  //
  //   await expect(
  //     traceFileDependencies(entryPoint, { workingDirectory: __dirname }),
  //   ).resolves.toMatchSnapshot();
  // });

  test('traceDependencies', async () => {
    const base = new URL('./fixtures/fixture1/', import.meta.url);

    const manifest = await loadJson<PackageJson>(new URL('package.json', base));

    await expect(
      traceDependencies(manifest.dependencies || {}, base).catch(logNThrow),
    ).resolves.toMatchSnapshot();
  });
});

// describe('Deps API', () => {
//   test('collect', async () => {});
//   test('flatten', async () => {});
//
//   test('traceFileDeps', async () => {
//     const entryPoint = require.resolve(
//       '../../api/dist/src/handlers/grpc/server.js',
//     );
//
//     await expect(
//       traceFileDependencies(entryPoint, { workingDirectory: __dirname }),
//     ).resolves.toMatchSnapshot();
//   });
//
//   test('listDeps', async () => {
//     const base = require.resolve('../../api/package.json');
//     const manifest = require(base);
//
//     await expect(
//       traceDependencies(manifest.dependencies, dirname(base)),
//     ).resolves.toMatchSnapshot();
//   });
// });

// describe('resolve-paths', () => {});
