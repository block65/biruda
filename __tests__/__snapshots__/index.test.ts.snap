// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Deps Fixtures traceDependencies 1`] = `
Array [
  Object {
    "files": undefined,
    "name": "aws-sdk",
    "path": "file:///home/mholman/Projects/block65/biruda/__tests__/fixtures/fixture1/node_modules/aws-sdk/",
    "version": "2.831.0",
  },
  Object {
    "files": undefined,
    "name": "buffer",
    "path": "file:///home/mholman/Projects/block65/biruda/__tests__/fixtures/fixture1/node_modules/buffer/",
    "version": "4.9.2",
  },
  Object {
    "files": undefined,
    "name": "base64-js",
    "path": "file:///home/mholman/Projects/block65/biruda/__tests__/fixtures/fixture1/node_modules/base64-js/",
    "version": "^1.0.2",
  },
  Object {
    "files": undefined,
    "name": "ieee754",
    "path": "file:///home/mholman/Projects/block65/biruda/__tests__/fixtures/fixture1/node_modules/buffer/node_modules/ieee754/",
    "version": "^1.1.4",
  },
  Object {
    "files": undefined,
    "name": "isarray",
    "path": "file:///home/mholman/Projects/block65/biruda/__tests__/fixtures/fixture1/node_modules/isarray/",
    "version": "^1.0.0",
  },
  Object {
    "files": undefined,
    "name": "events",
    "path": "file:///home/mholman/Projects/block65/biruda/__tests__/fixtures/fixture1/node_modules/events/",
    "version": "1.1.1",
  },
  Object {
    "files": undefined,
    "name": "ieee754",
    "path": "file:///home/mholman/Projects/block65/biruda/__tests__/fixtures/fixture1/node_modules/ieee754/",
    "version": "1.1.13",
  },
  Object {
    "files": undefined,
    "name": "jmespath",
    "path": "file:///home/mholman/Projects/block65/biruda/__tests__/fixtures/fixture1/node_modules/jmespath/",
    "version": "0.15.0",
  },
  Object {
    "files": undefined,
    "name": "querystring",
    "path": "file:///home/mholman/Projects/block65/biruda/__tests__/fixtures/fixture1/node_modules/querystring/",
    "version": "0.2.0",
  },
  Object {
    "files": Array [
      "lib/sax.js",
      "LICENSE",
      "LICENSE-W3C.html",
      "README.md",
    ],
    "name": "sax",
    "path": "file:///home/mholman/Projects/block65/biruda/__tests__/fixtures/fixture1/node_modules/sax/",
    "version": "1.2.1",
  },
  Object {
    "files": undefined,
    "name": "url",
    "path": "file:///home/mholman/Projects/block65/biruda/__tests__/fixtures/fixture1/node_modules/url/",
    "version": "0.10.3",
  },
  Object {
    "files": Array [
      "LICENSE-MIT.txt",
      "punycode.js",
    ],
    "name": "punycode",
    "path": "file:///home/mholman/Projects/block65/biruda/__tests__/fixtures/fixture1/node_modules/punycode/",
    "version": "1.3.2",
  },
  Object {
    "files": undefined,
    "name": "querystring",
    "path": "file:///home/mholman/Projects/block65/biruda/__tests__/fixtures/fixture1/node_modules/querystring/",
    "version": "0.2.0",
  },
  Object {
    "files": undefined,
    "name": "uuid",
    "path": "file:///home/mholman/Projects/block65/biruda/__tests__/fixtures/fixture1/node_modules/uuid/",
    "version": "3.3.2",
  },
  Object {
    "files": Array [
      "lib",
    ],
    "name": "xml2js",
    "path": "file:///home/mholman/Projects/block65/biruda/__tests__/fixtures/fixture1/node_modules/xml2js/",
    "version": "0.4.19",
  },
  Object {
    "files": Array [
      "lib/sax.js",
      "LICENSE",
      "README.md",
    ],
    "name": "sax",
    "path": "file:///home/mholman/Projects/block65/biruda/__tests__/fixtures/fixture1/node_modules/xml2js/node_modules/sax/",
    "version": ">=0.6.0",
  },
  Object {
    "files": undefined,
    "name": "xmlbuilder",
    "path": "file:///home/mholman/Projects/block65/biruda/__tests__/fixtures/fixture1/node_modules/xmlbuilder/",
    "version": "~9.0.1",
  },
]
`;
